### PR TITLE
Add small header cache

### PR DESCRIPTION
**Pull request by arkpar**
This saves a few DB queries when syncing as the same hader is being queried multiple times.

### CODE REVIEW

File: client/db/src/lib.rs [hunk 1]
```diff
@@ -407,7 +422,24 @@ impl<Block: BlockT> BlockchainDb<Block> {
...
+				if let Some(result) = cache.get_refresh(h) {
+					return Ok(result.clone());
+				}
+				match utils::read_header(&*self.db, columns::KEY_LOOKUP, columns::HEADER, id) {
+					Ok(header) => {
+						cache_header(&mut cache, h.clone(), header.clone());
+						Ok(header)
+					}
+					e @Err(_) => e,
+				}
```
Comment (bkchr) [inline client/db/src/lib.rs:437]:
```suggestion
				let header = utils::read_header(&*self.db, columns::KEY_LOOKUP, columns::HEADER, id)?;
				
				cache_header(&mut cache, h.clone(), header.clone());
				Ok(header)
```

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -352,12 +354,24 @@ impl<'a> sc_state_db::MetaDb for StateMetaDb<'a> {
 	}
 }
 
+fn cache_header<Hash: std::cmp::Eq + std::hash::Hash, Header>(
+	cache: &mut LinkedHashMap<Hash, Option<Header>>,
+	hash: Hash,
+	header: Option<Header>
```
Comment (bkchr) [inline client/db/src/lib.rs:360]:
```suggestion
	header: Option<Header>,
```

File: client/db/src/lib.rs [hunk 3]
```diff
@@ -424,12 +456,17 @@ impl<Block: BlockT> sc_client_api::blockchain::HeaderBackend<Block> for Blockcha
...
-			BlockId::Hash(_) => read_db(
-				&*self.db,
-				columns::KEY_LOOKUP,
-				columns::HEADER,
-				id
-			)?.is_some(),
+			BlockId::Hash(h) => match self.header_cache.lock().get(&h) {
+				Some(header) => header.is_some(),
+				None => {
+					read_db(
```
Comment (bkchr) [inline client/db/src/lib.rs:462]: Why don't we cache here?

**Review by bkchr — APPROVED** [review]

File: client/db/src/lib.rs (hunk 3 above)
Comment (arkpar) [reply client/db/src/lib.rs]: Changed it to simply use `header()`

**Comment by gnunicorn** [comment]
@arkpar are you still working on this? The polkadot companion is still failing.

**Comment by arkpar** [comment]
Should be good to merge now

**Review by andresilva — APPROVED** [review]

File: client/db/src/lib.rs [hunk 4]
```diff
@@ -1306,6 +1327,12 @@ impl<Block: BlockT> Backend<Block> {
 			is_best,
 			mut cache,
 		)) = imported {
+			let header_metadata = CachedHeaderMetadata::from(&header);
+			self.blockchain.insert_header_metadata(
```
Comment (andresilva) [inline client/db/src/lib.rs:1331]: Good catch.

**Comment by andresilva** [comment]
bot merge

**Pull request merged**